core/src/smartcontracts/isi/permissions.rs

```diff
@@ -20,65 +23,362 @@ impl NeedsPermission for QueryBox {}
+        another: ValidatorType,
+    ) -> std::result::Result<(), ValidatorTypeMismatch> {
+        if self != another {
+            return Err(ValidatorTypeMismatch::expected(another).found(self));
+        }
+
+        Ok(())
+    }
+}
+
+impl std::fmt::Display for ValidatorType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ValidatorType::Instruction => write!(f, "Instruction"),
+            ValidatorType::Query => write!(f, "Query"),
+            ValidatorType::Expression => write!(f, "Expression"),
+        }
+    }
+}
+
+pub mod error {
+    //! Contains errors structures
+
+    use super::ValidatorType;
+
+    /// Reason for prohibiting the execution of the particular instruction.
+    #[derive(Debug, Clone, thiserror::Error)]
+    #[allow(variant_size_differences)]
+    pub enum DenialReason {
+        /// [`ValidatorTypeMismatch`] variant
+        #[error("{0}")]
+        ValidatorTypeMismatch(#[from] ValidatorTypeMismatch),
+        /// Variant for custom error
+        #[error("{0}")]
+        Custom(String),
+        /// Variant used when at least one [`Validator`](super::IsAllowed) should be provided
+        #[error("No validators provided")]
+        NoValidatorsProvided,
+    }
+
+    impl From<String> for DenialReason {
+        fn from(s: String) -> Self {
+            Self::Custom(s)
+        }
+    }
+
+    /// Wrong validator expectation error
+    ///
+    /// I.e. used when user tries to validate [`QueryBox`](super::QueryBox) with
+    /// [`IsAllowedBoxed`](super::IsAllowedBoxed) containing
+    /// [`IsAllowedBoxed::Instruction`](super::IsAllowedBoxed::Instruction) variant
+    #[derive(Debug, Copy, Clone, thiserror::Error)]
+    #[error("Expected `{expected}` validator type, but found `{found}`")]
+    pub struct ValidatorTypeMismatch {
+        expected: ValidatorType,
+        found: ValidatorType,
+    }
+
+    impl ValidatorTypeMismatch {
+        /// Start construction of [`ValidatorTypeMismatch`] by providing *expected* type
+        pub fn expected(expected: ValidatorType) -> Expected {
+            Expected { expected }
+        }
+    }
+
+    /// Helper struct for constructing [`ValidatorTypeMismatch`]
+    #[derive(Debug, Copy, Clone)]
+    pub struct Expected {
+        expected: ValidatorType,
+    }
+
+    impl Expected {
+        /// Finish construction of [`ValidatorTypeMismatch`] by providing *found* type
+        pub fn found(self, found: ValidatorType) -> ValidatorTypeMismatch {
+            ValidatorTypeMismatch {
+                expected: self.expected,
+                found,
+            }
+        }
+    }
+}
+
+use error::*;
+
+/// Result type for permission validators
+pub type Result<T> = std::result::Result<T, DenialReason>;
 
 /// Implement this to provide custom permission checks for the Iroha based blockchain.
-pub trait IsAllowed<W: WorldTrait, O: NeedsPermission> {
+pub trait IsAllowed<W: WorldTrait, O: NeedsPermission>:
```

Comment:
!!! **Most important part of the PR**

Reply:
Yeah, it adds `Debug`, `Clone` and `Serialize` bounds. Not sure if we need it now, but it can be useful later